Display notification filter/clear actions as menu items

nikclayton opened the pull request:
Previously the notification filter and clear actions were shown as buttons in the UI, with a preference that determined whether they were displayed.

Remove this preference, and display them as menu items.

- "Filter notifications" is shown as an icon, if possible
- "Clear notifications" is only ever shown as a menu item, to reduce the chance the user inadvertently selects it

To ensure that the options menu appears correctly, remove the code that creates a "fake" action bar, and adjust the layouts so that there are three toolbars;

- mainToolbar -- displays the icons, and the current "location" (Home, Notifications, etc)
- topNav -- displays the row of tabs at the top
- bottomNav -- displays the row of tabs at the bottom

Only one of them is set as the support action bar (depending on the user's preferences). This provides the "show a logo" and "show the options menu" functionality as standard, without needing to re-implement as the previous code did.

nikclayton commented:
Some screenshots of the new UI.

Here's with the topbar enabled, showing (a) the new filter shortcut icon, and (b) the options menu when it's open.

<img src="https://github.com/tuskyapp/Tusky/assets/773100/64d4f996-cb90-4a9c-972b-f33efd114b69" width="45%"> <img src="https://github.com/tuskyapp/Tusky/assets/773100/50d1c35f-ab65-43ea-a150-4b8f736e541e" width="45%">

Here's what it looks like with the topbar disabled:

<img src="https://github.com/tuskyapp/Tusky/assets/773100/63027c28-f2a8-452b-8923-01848497de9e" width="45%"> <img src="https://github.com/tuskyapp/Tusky/assets/773100/b454bd4d-65bf-4288-88c6-30253329ce9e" width="45%">

Same thing, but with the navigation at the bottom:

<img src="https://github.com/tuskyapp/Tusky/assets/773100/520cd0a3-f8f0-4663-97a1-b12ecfca231f" width="45%"> <img src="https://github.com/tuskyapp/Tusky/assets/773100/b6742a78-bb83-45de-b4b9-71425a593106" width="45%">

charlag commented:
I am not sure how I feel about options menu together with tabs but I support moving actions into toolbar & dropdown menu

nikclayton commented:
> I am not sure how I feel about options menu together with tabs but I support moving actions into toolbar & dropdown menu

The menu is required (it should have been added when removing the top nav bar was added as an option) because you need a place to put functions that would otherwise be gesture-only.

Otherwise those functions are not accessible to eg screen reader users.

Lakoja commented on app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationsFragment.kt:
```diff
@@ -369,10 +350,10 @@ class NotificationsFragment :
-                launch {
-                    viewModel.uiState.collectLatest { updateFilterVisibility(it.showFilterOptions) }
-                }
+                // Collec the uiState. Nothing is done with it, but if you don't collect it then
```
```suggestion
                // Collect the uiState. Nothing is done with it, but if you don't collect it then
```

Lakoja commented:
> Remove this preference, and display them as menu items.

I guess these changes need an issue?
There's a graphical / usage change that might need to be communicated.
And one might reference that issue in the settings migration comment (in `TuskyApplication`).

Lakoja commented on app/src/main/java/com/keylesspalace/tusky/MainActivity.kt:
```diff
@@ -257,6 +261,23 @@ class MainActivity : BottomSheetActivity(), ActionButtonActivity, HasAndroidInje
         val hideTopToolbar = preferences.getBoolean(PrefKeys.HIDE_TOP_TOOLBAR, false)
         binding.mainToolbar.visible(!hideTopToolbar)
 
+        // Configure a faux-options menu if the top toolbar is hidden
```
What is "faux" here?

nikclayton replied:
Fake / artificial.

Lakoja commented on app/src/main/java/com/keylesspalace/tusky/MainActivity.kt:
```diff
@@ -257,6 +261,23 @@ class MainActivity : BottomSheetActivity(), ActionButtonActivity, HasAndroidInje
+        if (hideTopToolbar) {
+            val navOnBottom = preferences.getString(PrefKeys.MAIN_NAV_POSITION, "top") == "bottom"
+            val view = if (navOnBottom) binding.bottomNavOptionsMenu else binding.topNavOptionsMenu
+            val popup = PopupMenu(this, view)
```
There is no "system" way of doing this?

(When the title is hidden there is no context menu?)

This fixes then another issue: If at the moment the title is hidden, "Refresh" and "Search" are not accessible.

It would be quite a bit deal here then.
I would move this code to a method then, something like `if (hideTopToolbar) emulateContextMenuForNavBar()`.

Lakoja commented:
On my emulator (Android 12) it looks like this at the moment:
![grafik](https://github.com/tuskyapp/Tusky/assets/1618905/e942d79d-1b98-492b-9848-b89bf5cfaf51)

So two problems: 
- It has no different background color and is thus melting with the timeline content.
- The wording is a bit inconsistent.
(Search is general, the other 4 are for the current list, consistently it should probably be called "Load newest" then

Both could be tackled in another issue/PR. However the first one might also be solved here.

nikclayton commented:
@Lakoja @charlag 

Turns out the code before this PR was overly complicated, and it's straightforward to have three toolbars (one for the mainnav, one for the top tabs, and one for the bottom tabs).

Whichever one of those is assigned to be the support action bar will display the logo and the options menu "for free", and does it with the correct theme/colours (which addresses one of your other comments).

I've adjusted the UI strings to make it clear these are operations on notifications as well.

I've updated the screenshots in the first comment on this PR to show the current behaviour.

PTAL.

Lakoja commented on app/src/main/java/com/keylesspalace/tusky/MainActivity.kt:
```diff
@@ -350,6 +366,14 @@ class MainActivity : BottomSheetActivity(), ActionButtonActivity, HasAndroidInje
+    override fun onPrepareMenu(menu: Menu) {
+        super.onPrepareMenu(menu)
+
+        // If the main toolbar is hidden then there's no space in the top/bottomNav to show
```
Just as an observation / question / other issue: There might be enough room on larger devices (tablets, landscape phones)?

nikclayton replied:
There might be, but:

a. It's difficult to reliably tell.
b. The UX is confusing if a single row contains tab icons and action icons.

I'll take a screenshot of that so you can see what I mean.

nikclayton replied:
Both of these are **without** the `onPrepareMenu` code in this PR.

Portrait on a phone:

![image](https://github.com/tuskyapp/Tusky/assets/773100/fe99b876-66cd-44a1-a5c0-a71c452bfd6d)

Portrait on a tablet:

![image](https://github.com/tuskyapp/Tusky/assets/773100/a380c96f-1eb3-45aa-9226-d5e912f50ba8)

If you pay careful attention to the colour and alignment of the search icon it's possible to tell that it's different, but not really at first glance.

Lakoja replied:
More or less what I would have expected from the tablet view: quite unfortunate (the spacing of the icons).

(But still a different issue.)

Lakoja commented:
> Whichever one of those is assigned to be the support action bar will display the logo and the options menu "for free", and does it with the correct theme/colours (which addresses one of your other comments).

Very nice!

> PTAL.

PTAL? :-)

One of my comments is not yet resolved: This solves a rather noticeable bug (context menu not reachable) which should have an issue, I guess.

nikclayton commented:
> > PTAL.
> 
> PTAL? :-)

Please take a look.

> One of my comments is not yet resolved: This solves a rather noticeable bug (context menu not reachable) which should have an issue, I guess.

I don't understand what you mean here.

Lakoja commented:
> > > PTAL.
> > 
> > 
> > PTAL? :-)
> 
> Please take a look.
> 
> > One of my comments is not yet resolved: This solves a rather noticeable bug (context menu not reachable) which should have an issue, I guess.
> 
> I don't understand what you mean here.

I gather then that we don't want to have issues for meaninful changes (for documentation / communication / release notes / ...)?

Here in this PR we have two quite distinct changes:
- Notification filter button goes to context menu
- Context menu is shown also when title is hidden

I have no reservations against doing both but I _would_ document either change as an issue.

nikclayton commented:
> I gather then that we don't want to have issues for meaninful changes (for documentation / communication / release notes / ...)?

Not always, no. It can help, but it's not mandatory.

That's why it's important that the PR description stand alone in describing what has changed and why. If you look at recent PRs that I've merged from other people you'll see that I sometimes rewrite the PR description and/or title to try and make what's changed and why easier to understand.

Release notes are written by generating a list of the merged PRs since the previous tag (e.g., https://github.com/tuskyapp/Tusky/compare/v23.0...develop) and reviewing each one to decide what to include in the release notes.
